### PR TITLE
Change strategy of output.

### PR DESCRIPTION
**Pull request by yageek**
Use a different strategy to retrieve the content of STDOUT and STDERR.
The `ShellOutError` structure has been changed but the API remains the same.
The structure now contains both STDOUT and STDERR as `Data` and we can
retrieve the equivalent `String` using the previous properties `message`
and `output`.
From my memories, any non 0 `terminationStatus` should be considered as
an error.
This also allows to introduce a parameter `printOutput` which will
output the received data to `STDOUT` and `STDERR`.

### CODE REVIEW

**Comment by JohnSundell** [comment]
Hi @yageek! Thanks for this PR! Could you please outline what the target usecase is for this change - that is - why is it better than the current implementation? 🙂

**Comment by yageek** [comment]
The thing I wanted to add is still not working after some tests (last bullet) :(. The elements I wanted to improve:

- [X] Any non 0 output should be considered as an error, not only when something has been written to STDERR.
- [X] Possibility to retrieve any `Data` and the termination status when an `ShellOutError` is thrown.
- [X] As soon as characters are received by the `Process`, I would like to stream them directly out instead of waiting the completion to finish. In a CI script, it could be useful to stream the received elements as long as the compilation is going forward.

I am investigating a little further about how to stream the output correctly. Is there some slack channel to discuss about improvement and features but the way?

**Comment by yageek** [comment]
After taking a look more closely, this is working in deed. I have a created a compilation script that behaves like travis and I am able to see the output as it is streamed. This behaviour could be used also for `marathon run`.

File: Sources/ShellOut.swift [hunk 1]
```diff
@@ -23,81 +24,114 @@ import Foundation
...
+@discardableResult public func shellOut(to commands: [String], at path: String = ".", printOutput: Bool = false) throws -> String {
     let command = commands.joined(separator: " && ")
-    return try shellOut(to: command, at: path)
+    return try shellOut(to: command, at: path, printOutput: printOutput)
 }
 
 // Error type thrown by the `shellOut()` function, in case the given command failed
 public struct ShellOutError: Swift.Error {
+    /// The buffer data retuned by `STDERR`.
+    public let stderr: Data
```
Comment (JohnSundell) [inline Sources/ShellOut.swift:57]: Why is this needed, since we already have `message` and `output`?

File: Sources/ShellOut.swift [hunk 2]
```diff
@@ -23,81 +24,114 @@ import Foundation
  */
 @discardableResult public func shellOut(to command: String,
                                         arguments: [String] = [],
-                                        at path: String = ".") throws -> String {
+                                        at path: String = ".",
+                                        printOutput: Bool = false) throws -> String {
     let process = Process()
     let command = "cd \"\(path)\" && \(command) \(arguments.joined(separator: " "))"
-    return try process.launchBash(with: command)
+    return try process.launchBash(with: command, printOutput: printOutput)
```
Comment (JohnSundell) [inline Sources/ShellOut.swift]: How about injecting a `FileHandle` here instead? That way the API user can decide where to stream to, and you can use that to write tests against this new functionality.

**Comment by JohnSundell** [comment]
@yageek Thanks again for this, I love the idea - it's going to make a big improvement to Marathon, since you don't have to wait for the full output 👍 I put some comments inline, and also some tests are needed before this can be merged in 🙂

File: Sources/ShellOut.swift (hunk 2 above)
Comment (yageek) [reply Sources/ShellOut.swift]: Good point. I will change it.

**EDIT** What do you think about:
```swift
 public func shellOut(to commands: [String], at path: String = ".", redirectStdout: FileHandle? = nil, redirectStderr: FileHandle? = nil) throws -> String
```

Comment (JohnSundell) [reply Sources/ShellOut.swift]: 👍

File: Sources/ShellOut.swift (hunk 1 above)
Comment (yageek) [reply Sources/ShellOut.swift:57]: I would imagine some output should just be some bytes and are not intended to be represented as `string`.

Comment (JohnSundell) [reply Sources/ShellOut.swift:57]: But `string` is so much easier to consume, especially since this package is mostly targeted towards scripting, where convenience is key.

**Comment by yageek** [comment]
I have updated the API and added the missing tests. The only question remaining is to allow/disallow access to the raw `Data` in case of error.

File: Sources/ShellOut.swift [hunk 3]
```diff
@@ -23,81 +25,126 @@ import Foundation
  */
 @discardableResult public func shellOut(to command: String,
                                         arguments: [String] = [],
-                                        at path: String = ".") throws -> String {
+                                        at path: String = ".",
+                                       redirectStdout: FileHandle? = nil, redirectStderr: FileHandle? = nil) throws -> String {
```
Comment (JohnSundell) [inline Sources/ShellOut.swift]: How about calling this `outputHandle`, and the error one `errorHandle`?

**Review by JohnSundell — COMMENTED** [review]
Looks really good 👍 Still not sure about adding the `Data` properties to the error type, why are they really needed, since the error already provides all other info you should need? (The termination status is a great addition though!)

**Comment by yageek** [comment]
I renamed the arguments. Your proposition made more sens. About the error, according to me, we should give the opportunity to have access to the data. Here when we are converting to `String`we assume we are using an `utf8` encoding. May be some programs outputs text in a different encoding. Providing an access to the internal buffers will allow any transformation.

**Review by JohnSundell — APPROVED** [review]
Awesome! Thanks @yageek! 🚀
